src/hotspot/share/opto/memnode.cpp

```diff
@@ -3519,17 +3519,130 @@ Node *StoreNode::Ideal_masked_input(PhaseGVN *phase, uint mask) {
-      Node *shl = val->in(1);
-      if( shl->Opcode() == Op_LShiftI ) {
-        const TypeInt *t2 = phase->type( shl->in(2) )->isa_int();
-        if( t2 && t2->is_con() && (t2->get_con() == t->get_con()) ) {
-          set_req_X(MemNode::ValueIn, shl->in(1), phase);
-          return this;
+// If (conIL > conIR) we are inventing 0 lower bits, and throwing
+// away upper bits, but we are not introducing garbage bits from the upper bits
+// (the right in the notation below).
+// We can simplify into
+// (StoreB ... (LShiftI _ valIn (conIL - conIR)) )
+// This case happens when the right-hand side of the store was itself a left shift, that gets merged
+// into the inner left shift of the sign-extension. For instance, if we have
+// array_of_shorts[0] = (short)(X << 2)
+// We get a structure such as:
+// (StoreB ... (RShiftI _ (LShiftI _ (LShiftI _ X 2) 16) 16))
+// that is simplified into
+// (StoreB ... (RShiftI _ (LShiftI _ X 18) 16)).
+// It is thus useful to handle the case where conIL > conIR.
+//
+// Let's assume we have the following 32 bits integer that we want to stuff in 8 bits char:
+// +------------------------+---------+
+// |        v[8..31]        | v[0..7] |
+// +------------------------+---------+
+//  31                     8 7        0
+// v[0..7] is meaningful, but v[8..31] is not. In this case, num_rejected_bits == 24.
+// Let's study what happens in different cases to see that the simplification into
+// (StoreB ... (LShiftI _ valIn (conIL - conIR)) )
+// is valid if:
+// - conIL >= conIR
+// - conIR <= num_rejected_bits
+// Let's also remember that conIL < 32 since (x << 33) is simplified into (x << 1)
+// and (x << 31) << 2 is simplified into 0. This means that in any case, after the
+// left shift, we always have at least one bit of the original v.
+//
+// ### Case 1 : conIL == conIR
+// If we do the shift left then right by 24 bits, we get:
+// after << 24
+// +---------+------------------------+
+// | v[0..7] |           0            |
+// +---------+------------------------+
+//  31     24 23                      0
+// after >> 24
+// +------------------------+---------+
+// |        sign bit        | v[0..7] |
+// +------------------------+---------+
+//  31                     8 7        0
+// The non-rejected bits are the same before and after, so, indeed, simplifying is fine.
+//
+// ### Case 2: conIL > conIR == num_rejected_bits
+// We take conIL == 26 for this example.
+// after << 26
+// +---------+------------------------+
+// | v[0..5] |           0            |
+// +---------+------------------------+
+//  31     26 25                      0
+// after >> 24
+// +------------------+---------+-----+
+// |     sign bit     | v[0..5] |  0  |
+// +------------------+---------+-----+
+//  31               8 7       2 1   0
+// The non-rejected bits are the 8 lower one of (v << conIL - conIR).
+// The bits 6 and 7 of v have been thrown away after the shift left.
+// The simplification is still fine.
```

Comment:
Suggestion for everywhere: `fine` -> `valid`.

Reply:
Or correct.

Reply:
I like correct.